domain/user/state/state.go

```diff
@@ -0,0 +1,395 @@
+		}
+		return errors.Trace(err)
+	})
+}
+
+// SetActivationKey removes any active passwords for the user and sets the
+// activation key. If no user is found for the supplied name an error
+// is returned that satisfies usererrors.NotFound.
+func (st *State) SetActivationKey(ctx context.Context, uuid user.UUID, activationKey string) error {
+	err := uuid.Validate()
+	if err != nil {
+		return fmt.Errorf("cannot validate user uuid %q: %w", uuid, err)
+	}
+
+	removed, err := st.isRemoved(ctx, uuid)
+	if err != nil {
+		return fmt.Errorf("cannot get user %q: %w", uuid, err)
+	}
+	if removed {
+		return fmt.Errorf("cannot set activation key for removed user %q", uuid)
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	return db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		err = removePasswordHash(ctx, tx, uuid)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return setActivationKey(ctx, tx, uuid, activationKey)
+	})
+}
+
+// SetPasswordHash removes any active activation keys and sets the user
+// password hash and salt. If no user is found for the supplied name an error
+// is returned that satisfies usererrors.NotFound.
+func (st *State) SetPasswordHash(ctx context.Context, uuid user.UUID, passwordHash string, salt []byte) error {
+	err := uuid.Validate()
+	if err != nil {
+		return fmt.Errorf("cannot validate user uuid %q: %w", uuid, err)
+	}
+
+	removed, err := st.isRemoved(ctx, uuid)
+	if err != nil {
+		return fmt.Errorf("cannot get user %q: %w", uuid, err)
+	}
+	if removed {
+		return fmt.Errorf("cannot set password hash for removed user %q", uuid)
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	return db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		err = removeActivationKey(ctx, tx, uuid)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return setPasswordHash(ctx, tx, uuid, passwordHash, salt)
+	})
+}
+
+func addUser(ctx context.Context, tx *sql.Tx, uuid user.UUID, user user.User, creatorUuid user.UUID) error {
+	addUserQuery := "INSERT INTO user (uuid, name, display_name, created_by_uuid, created_at) VALUES (?, ?, ?, ?, ?)"
+
+	_, err := tx.ExecContext(ctx, addUserQuery,
+		uuid.String(),
+		user.Name,
+		user.DisplayName,
+		creatorUuid.String(),
+		time.Now(),
+	)
+	if databaseutils.IsErrConstraintUnique(err) {
+		return usererrors.AlreadyExists
+	} else if databaseutils.IsErrConstraintForeignKey(err) {
+		return usererrors.UserCreatorUuidNotFound
+	} else if err != nil {
+		return fmt.Errorf("cannot insert user %q: %w", user.Name, err)
+	}
+
+	return nil
+}
+
+func setPasswordHash(ctx context.Context, tx *sql.Tx, uuid user.UUID, passwordHash string, salt []byte) error {
+	defineUserAuthenticationQuery := `
+INSERT INTO user_authentication (user_uuid, disabled) 
+VALUES (?, ?) 
+ON CONFLICT(user_uuid) DO UPDATE SET user_uuid = excluded.user_uuid, disabled = excluded.disabled
```

Comment:
This is introducing a security problem for us because we potentially be setting a disabled user back to enabled.

Update the sql to just be `ON CONFLICT DO NOTHING`

Reply:
I am thinking about this a little more and I think this is the SQL we want.....

```sql
INSERT INTO user_authentication (user_uuid, disabled) VALUES (?, ?)
ON CONFLICT(user_uuid) DO UPDATE SET user_uuid = excluded.user_uuid
WHERE disabled = false
```

The reason for this is we want to see a side affect from the database to tell us that the user is not removed.

So on ExecContext we can check the result for the number of rows affected and if it is not 1 then we know the user auth is currently disabled and that we should not be allowing the password change to continue.

Hope this makes sense.

Reply:
make sense!  Thx for this catch!